### PR TITLE
Fragment generation & Color constraint adherence

### PR DESCRIPTION
**Pull request by yen-mai-codes**
Added Constraint class. Artwork can set constraint:
- color constraint
- line type constraint

Added Fragment class. Fragment class can now generate fragment with the following properties:
- randomness
- constraint adherence
- artwork's dimensions adherence

Future work:
- line type adherence
- non-random fragment

### CODE REVIEW

**Comment by nghi01** [comment]
Added Merging Function. The current merging function in peer class will merge one fragment at a time, but we can change that.

Future work:
- Refactoring of code (more get, set function for encapsulation)

File: src/main/py/commission/artfragment.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            coordinates = (
+                randrange(x_coordinate, x_bound),
+                randrange(y_coordinate, y_bound),
+            )
+
+            pixel_info = (coordinates, color_constraint)
+            fragment.add(pixel_info)
+            num_of_pixels -= 1
+
+        return fragment
```
Comment (zimmermatt) [inline src/main/py/commission/artfragment.py:78]: This method does quite a lot of somewhat tricky calculations. I would like to see it broken up by pulling those calcs into small functions. This will make it so we can have isolated tests for those discrete calcs and the overall test to make sure they're working together as expected.

File: src/main/py/commission/artfragment.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        )
+
+        # generate the set of pixels to occupy
+        num_of_pixels = randrange(0, width * height)
+        fragment = set()
+
+        # generate set of coordinates with color constraint adherence.
+        color_constraint = artwork.get_constraint().get_color()
+        x_bound = x_coordinate + width
+        y_bound = y_coordinate + height
```
Comment (zimmermatt) [inline src/main/py/commission/artfragment.py:66]: A thought... `width`/`height` seem to only used to set the number of pixels above and the x/y bound here.

* Should the class attribute be `x_bound`/`y_bound` instead?
* Are either actually accessed outside of this method?
* If not, maybe they can both be local to this method.

File: src/main/py/commission/constraint.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+class Constraint:
+    """
+    Class to manage constraint
+    """
+
+    def __init__(self, color: tuple[int, int, int, int], line_type: string):
+        """
+        Initializes an instance of the Constraint class.
+        - color (tuple(int, int, int, int)): color constraint.
```
Comment (zimmermatt) [inline src/main/py/commission/constraint.py:19]: Please doc the meaning of the 4-tuple fields. Better yet, also use a [namedtuple](https://docs.python.org/3/library/collections.html#collections.namedtuple).

**Review by zimmermatt — CHANGES_REQUESTED** [review]

File: src/main/py/commission/artfragment.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        TODO:
+        - implement address-based adherence
+        - implement line type adherence
+        """
+        # generate (x,y) coordinates and width, height
+        x_coordinate = randrange(0, artwork.width + 1)
+        y_coordinate = randrange(0, artwork.height + 1)
+
+        width = randrange(x_coordinate, artwork.width + 1) - x_coordinate + 1
+        height = randrange(y_coordinate, artwork.height + 1) - y_coordinate + 1
```
[the remark this answers is not in the document]
Comment (zimmermatt) [reply src/main/py/commission/artfragment.py:47]: A revision I made to this comment didn't take... the last comment isn't what I meant to send. I'll restate it in a follow-up momentarily.

[the remark this answers is not in the document]
Comment (zimmermatt) [reply src/main/py/commission/artfragment.py:47]: The last sentence should read:

If the intent here is to guarantee a non-zero result while ensuring bounds are not exceeded, then
```
  width  = randrange(1, 2 + artwork.width - x_coordinate))
  height = randrange(1, 2 + artwork.height - y_coordinate))
```
would be the way to go.

**Comment by zimmermatt** [comment]
> Added Merging Function. The current merging function in peer class will merge one fragment at a time, but we can change that.
> 
> Future work:
> 
> * Refactoring of code (more get, set function for encapsulation)

It's okay this time... don't want to hold things up, so no need to split it back out. However, ideally fragment gen and merging would be separate commits/PRs, since they're discrete bits of functionality (though I recognize there's some dependencies).

File: src/main/py/commission/artfragment.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        - contributor (string): Peer ID that created the artfragment.
+        - artwork (Artwork): artwork that the fragment is intended for.
+        - fragment (set): list of (x,y,color) based on constraint.
+        """
+        self.width = 0
+        self.height = 0
+        self.coordinates = ()
+        self.contributor = contributor
+        self.artwork = artwork
+        self.fragment = self.generate_fragment(artwork)
```
[the remark this answers is not in the document]
Comment (yen-mai-codes) [reply src/main/py/commission/artfragment.py:33]: note to self:

for peerA's art fragment generation flow:
1. peerA calls `generator = ArtFragmentGenerator(peerA, artworkA)`
2. peerA additionally calls `fragment = generator.generate_fragment()` which returns an instance of immutable dataclass `ArtFragment`

File: src/main/py/commission/artfragment.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zimmermatt) [reply src/main/py/commission/artfragment.py]: Correction for how much to add from our conversation (good catch :-))... I had `width`/`height` mixed up in my head as `0` based, so added `2` (D'OH!). But you're right, we should just add `1`.

```
  width  = randrange(1, 1 + artwork.width - x_coordinate))
  height = randrange(1, 1 + artwork.height - y_coordinate))
```

Looking at this again, there's one additional thing to be aware of that doesn't apply in Python (numerics are arbitrary precision). In cases where numeric primitives are fixed precision, one needs to be mindful of overflow. Two ways to protect against this are:
1. Do the addition last
1. Use a function that is inclusive on both the lower and upper bounds, like `randint()` in Python

So the above becomes:

```
  width  = randrange(1, artwork.width - x_coordinate + 1))
  height = randrange(1, artwork.height - y_coordinate + 1))
```

File: src/main/py/drawing/color.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        """
+        Create a valid color tuple based on input
+        """
+        normalized_color = []
+        for channel in channels:
+            normalized_channel = self.normalize_chanel(channel)
+            normalized_color.append(normalized_channel)
+        return tuple(normalized_color)
+
+    def normalize_chanel(self, channel):
```
Comment (zimmermatt) [inline src/main/py/drawing/color.py:34]: `s/chanel/channel/`

File: src/main/py/drawing/color.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        """
+        If a channel < 0, make it 0. If it is > 255, make it 255.
+        """
+        if channel < 0:
+            channel = 0
+        elif channel > 255:
+            channel = 255
+        return channel
+
+    def get_channels(self):
```
Comment (zimmermatt) [inline src/main/py/drawing/color.py:44]: No real need for a getter in this case.

File: src/main/py/drawing/pixel.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    def __init__(self, coordinates: Coordinates, color: Color):
+        """
+        Initializes an instance of the Pixel class.
+        - coordinates (Coordinates): represents Pixel's location.
+        - color (Color): represents Pixel's color.
+        """
+        self.coordinates = coordinates
+        self.color = color
+
+    def get_coordinates(self):
```
Comment (zimmermatt) [inline src/main/py/drawing/pixel.py:26]: No need for getters.

File: src/main/py/drawing/pixel.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        """
+        Get coordianates of pixel.
```
Comment (zimmermatt) [inline src/main/py/drawing/pixel.py:28]: If this method is kept for some reason `s/coordianates/coordinates/`

File: src/test/py/artfragmentgenerator/artfragmentgenerator_test.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        x_coordinate = subcanvas[0][0]
+        y_coordinate = subcanvas[0][0]
+        subcanvas_width = subcanvas[1][0]
+        subcanvas_height = subcanvas[1][1]
+
+        width_bound = artwork.width - x_coordinate + 1
+        height_bound = artwork.height - y_coordinate + 1
+
+        self.assertLessEqual(x_coordinate, artwork.width)
+        self.assertLessEqual(y_coordinate, artwork.height)
```
Comment (zimmermatt) [inline src/test/py/artfragmentgenerator/artfragmentgenerator_test.py:41]: x,y should only be `<` width,height

File: src/test/py/artfragmentgenerator/artfragmentgenerator_test.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        pixels = self.artfragment_generator.generate_pixels(
+            subcanvas, artwork.get_constraint()
+        )
+        color_constraint = artwork.get_constraint().get_color()
+
+        for pixel in pixels:
+            coordinates = pixel.coordinates
+            color = pixel.color
+            self.assertLessEqual(coordinates.x, artwork.width)
+            self.assertLessEqual(coordinates.y, artwork.height)
```
Comment (zimmermatt) [inline src/test/py/artfragmentgenerator/artfragmentgenerator_test.py:63]: x,y should only be `<` width,height

File: src/main/py/drawing/coordinates.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+    def normalize_coordinate(self, coordinate):
+        """
+        If coordinate < 0, make it 0.
+        """
+        if coordinate < 0:
+            return 0
+        return coordinate
+
+    def create_bounds(self, width, height):
```
Comment (zimmermatt) [inline src/main/py/drawing/coordinates.py:31]: I didn't see tests for this method. One should be added that tests the various boundary conditions, such as:
* x/y is the last pixel
* x/y is the next to last pixel
* x/y is at 0,0

**Review by zimmermatt — CHANGES_REQUESTED** [review]
Lots of great improvements in this. A few things to address, though.

**Comment by yen-mai-codes** [comment]
TODO: Story 3 - Update fragment generation to sign the completed fragment with the originator’s public key

**Review by zimmermatt — APPROVED** [review]
Mostly lgtm, so I'm going to commit it. There is one misspelling that should be fixed in a follow-up. The others can be opportunistically addressed.
